packages/templates/src/generators/sfdxGenerator.ts

```
@@ -29,4 +38,40 @@ export abstract class SfdxGenerator<
    * Validate provided options
    */
   public abstract validateOptions(): void;
+
+  /**
+   * Set source root to built-in templates or custom templates root if available.
+   * @param partialPath the relative path from the templates folder to templates root folder.
+   */
+  public sourceRootWithPartialPath(partialPath: string): void {
+    this.builtInTemplatesRootPath = path.join(
+      __dirname,
+      '..',
+      'templates',
+      partialPath
+    );
+    const { customTemplatesRootPath } = TemplateService.getInstance();
+    if (!customTemplatesRootPath) {
+      this.sourceRoot(path.join(this.builtInTemplatesRootPath));
+    } else {
+      if (fs.existsSync(path.join(customTemplatesRootPath, partialPath))) {
+        this.sourceRoot(path.join(customTemplatesRootPath, partialPath));
+      }
+    }
+  }
+
+  public templatePath(...paths: string[]): string {
+    // The template paths are relative to the generator's source root
+    // If we have set a custom template root, the source root should have already been set.
+    // Otherwise we'll fallback to the built-in templates
+    const customPath = super.templatePath(...paths);
+    if (fs.existsSync(customPath)) {
+      return customPath;
+    } else {
+      // files that are builtin and not in the custom template folder
+      return super.templatePath(
+        path.join(this.builtInTemplatesRootPath!, ...paths)
```

Is the non-null assertion necessary here? Is a null builtInTemplatesRootPath something we can test for?

---

we could throw an error about builtInTemplatesRootPath being null, but since that should always be set by the generator it shouldn't happen.